mantidproject/mantid
# SansILLReduction sensitivity calculation with horizontal offsets

**Description of work.**

Support has been added for multiple inputs for calculation of sensitivity maps in the SANS ILL workflow. This feature can be used to remove gaps in the sensitivity maps caused by the presence of beam stop that requires masking. If two measurements are done at different offsets, e.g. horizontal, then the same pixel may be masked only in one measurement and this information can be used to create sensitivity map without any gaps caused by beam stop masks. 

This special merging is done in `CalculateEfficiency` algorithm when a `WorkspaceGroup` is provided as `InputWorkspace` and the new boolean property, `MergeGroup`, is set. Currently, the efficiency calculation for the SANS at the ILL is called from `SANSILLReduction`, and no changes have been introduced to that algorithm. Instead, when a new property of the  `SANSILLAutoProcess`: `SensitivityWithOffsets` is set, then the output from the reference sample processing is grouped and then provided to `CalculateEfficiency` algorithm through `InputWorkspace` with  `MergeGroup`, set to `True`. The output of the  `CalculateEfficiency` will be then free of the beam-stop masks and can be used in further steps of data processing.
 
**To test:**

`SANSILLReduction`

1. In the workbench, make sure `Data/SystemTest/ILL/D22` is in your user directories.
2. Run the following script for 'manual' step-by-step reduction with two water reference measurements, measured with a horizontal offset:

```
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

# Load the mask
LoadNexusProcessed(Filename='D22_mask.nxs', OutputWorkspace='mask')
LoadNexusProcessed(Filename='D22_mask_central.nxs', OutputWorkspace='mask_central')
LoadNexusProcessed(Filename='D22_mask_offset.nxs', OutputWorkspace='mask_offset')

# Absorber
SANSILLReduction(Run='241238', ProcessAs='Absorber', OutputWorkspace='Cd')

# Beam
SANSILLReduction(Run='241226', ProcessAs='Beam', AbsorberInputWorkspace='Cd', 
                 OutputWorkspace='Db', FluxOutputWorkspace='fl')

# Container transmission known
CreateSingleValuedWorkspace(DataValue=0.94638, ErrorValue=0.0010425, OutputWorkspace='ctr')
AddSampleLog(Workspace='ctr', LogName='ProcessedAs', LogText='Transmission')

# Container
SANSILLReduction(Run='241239', ProcessAs='Container', AbsorberInputWorkspace='Cd', BeamInputWorkspace='Db',
                 TransmissionInputWorkspace='ctr', OutputWorkspace='can')

# Sample transmission known
CreateSingleValuedWorkspace(DataValue=0.52163, ErrorValue=0.00090538, OutputWorkspace='str')
AddSampleLog(Workspace='str', LogName='ProcessedAs', LogText='Transmission')

# Reference
SANSILLReduction(Run='344411', ProcessAs='Sample', MaskedInputWorkspace='mask_central',
                 AbsorberInputWorkspace='Cd', BeamInputWorkspace='Db', ContainerInputWorkspace='can',
                 OutputWorkspace='ref1', SensitivityOutputWorkspace='sens1')
SANSILLReduction(Run='344407', ProcessAs='Sample', MaskedInputWorkspace='mask_offset',
                 AbsorberInputWorkspace='Cd', BeamInputWorkspace='Db', ContainerInputWorkspace='can',
                 OutputWorkspace='ref2', SensitivityOutputWorkspace='sens2')

GroupWorkspaces(InputWorkspaces=['ref1', 'ref2'], OutputWorkspace='sensitivities')
CalculateEfficiency(InputWorkspaceGroup='sensitivities', OutputWorkspace='sens')

# Sample
SANSILLReduction(Run='241240', ProcessAs='Sample', AbsorberInputWorkspace='Cd', BeamInputWorkspace='Db',
                 TransmissionInputWorkspace='str', ContainerInputWorkspace='can', MaskedInputWorkspace='mask',
                 SensitivityInputWorkspace='sens', OutputWorkspace='sample', FluxInputWorkspace='fl')

# Integration
SANSILLIntegration(InputWorkspace='sample', OutputWorkspace='iq') 
```
3. Observe that entries in the `sensitivities` group have a masked region caused by masking of the beam stop, while the output of the `CalculateEfficiency`  mostly removes this effect.

`SANSILLAutoProcess`

1. In the workbench, make sure `Data/SystemTest/ILL/D22` is in your user directories.
2. Run the following minimalist reduction script with two water reference measurements, measured with a horizontal offset:

```
SANSILLAutoProcess(
    SampleRuns='344411,344407',
    MaskFiles='D22_mask_central.nxs,D22_mask_offset',
    SensitivityOutputWorkspace='sens',
    SampleThickness=0.1,
    OutputWorkspace='ref',
    SensitivityWithOffsets=True
)
```
3. Observe that the sensitivity workspace `sens` has undergone the treatment to remove the beam stop. Check if the output of the `SANSILLAutoProcess` and 'manual' running in the previous example provide comparable outcome.



Resolves #30291
Part of #26906

Release notes added.



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
